### PR TITLE
Npc movement refactor - regular mobs now move and (almost) try to attack you

### PR DESCRIPTION
**Pull request by mettaursp**
added BattleState
added MovementState
added TaskState
ported routine behavior over to the new state systems
updated AiState to use TaskState for timing & implemented multiple nodes

### CODE REVIEW

File: Maple2.Server.Game/Manager/Field/FieldManager.State.cs [hunk 1]
```diff
@@ -171,6 +172,10 @@ public partial class FieldManager {
         return fieldItem;
     }
 
+    public int RemoveMeNextLocalId() {
+        return NextLocalId();
+    }
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Manager/Field/FieldManager.State.cs:177]: If NextLocalId is private, there must be reason why, but you if you need for removeme, then just make it public? 🤔

File: Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/AiState.cs [hunk 2]
```diff
@@ -134,7 +193,32 @@ private struct StackEntry {
...
     }
 
-    void Push(Entry entry) {
+    private SkillMetadata? TryGetSkill(int idx, short levelOverride = 0) {
+        // idx starts at 1
+        if (idx < 1 || idx > actor.Skills.Length) {
+            actor.AppendDebugMessage($"{actor.Value.Metadata.Name}[{actor.Value.Id}]\n");
+            actor.AppendDebugMessage($"{AiMetadata!.Name}\n");
+            actor.AppendDebugMessage($"Invalid Skill Idx {idx}\n");
+
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/AiState.cs:202]: lets also use logger here if debugging is not on

File: Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/AiState.cs [hunk 3]
```diff
@@ -134,7 +193,32 @@ private struct StackEntry {
...
+        SkillMetadata? skill = actor.Skills[idx - 1];
+
+        if (skill is not null && levelOverride != 0) {
+            actor.Field.SkillMetadata.TryGet(skill.Id, levelOverride, out skill);
+        }
+
+        if (skill is null) {
+            actor.AppendDebugMessage($"{actor.Value.Metadata.Name}[{actor.Value.Id}]\n");
+            actor.AppendDebugMessage($"{AiMetadata!.Name}\n");
+            actor.AppendDebugMessage($"Missing Skill Idx {idx}\n");
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/AiState.cs:215]: same thing here

File: Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/AiState.cs [hunk 4]
```diff
@@ -413,19 +550,48 @@ private struct StackEntry {
...
+        return node.Value > health.Current / health.Total;
     }
 
     private bool ProcessCondition(StateCondition node) {
-        return false;
+        if (actor.BattleState.Target is null) {
+            return false;
+        }
+
+        Maple2.Model.Enum.ActorState state = Maple2.Model.Enum.ActorState.None;
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/AiState.cs]: can we use import here? if not its ok

File: Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/AnimationState.cs [hunk 5]
```diff
@@ -87,31 +92,61 @@ public struct TickPair {
 
         bool found = RigMetadata.Sequences.TryGetValue(name, out AnimationSequence? sequence);
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/AnimationState.cs:93]: I would rename this to `animationSequenceExists` or similar

File: Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/AnimationState.cs [hunk 6]
```diff
@@ -87,31 +92,61 @@ public struct TickPair {
...
-        ResetSequence();
+        if (reportingKeyframeEvent) {
+            queuedResetSequence = true;
+        } else {
+            if (PlayingSequence != null && actor is FieldNpc npc) {
+                npc.SendControl = true;
+            }
+
+            ResetSequence();
+        }
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/AnimationState.cs:149]: formatting might be off

```suggestion
        if (reportingKeyframeEvent) {
            queuedResetSequence = true;
            return;
        }
        if (PlayingSequence != null && actor is FieldNpc npc) {
            npc.SendControl = true;
        }

        ResetSequence();
```

File: Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/MovementState.Tasks.cs [hunk 7]
```diff
@@ -0,0 +1,322 @@
+﻿using Maple2.Model.Enum;
+using Maple2.Server.Game.Model.Enum;
+using Maple2.Server.Game.Model.Skill;
+using System.Numerics;
+using static Maple2.Server.Game.Model.Field.Actor.ActorStateComponent.TaskState;
+
+namespace Maple2.Server.Game.Model.Field.Actor.ActorStateComponent;
+
+public partial class MovementState {
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/MovementState.Tasks.cs:9]: I feel like these tasks classes could live on their own files, this file is big and could get bigger?

File: Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/MovementState.cs [hunk 8]
```diff
@@ -0,0 +1,571 @@
...
+
+                return;
+            } else {
+                float distance = (float) Math.Sqrt(distanceSquared);
+
+                walkDirection = (1 / distance) * offset;
+                tickDistance = Math.Min(tickDistance, distance);
+
+                actor.Transform.LookTo(walkDirection);
+            }
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/MovementState.cs:372]: remove else
```suggestion
            }
            float distance = (float) Math.Sqrt(distanceSquared);

            walkDirection = (1 / distance) * offset;
            tickDistance = Math.Min(tickDistance, distance);

            actor.Transform.LookTo(walkDirection);
```

File: Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs [hunk 9]
```diff
@@ -290,18 +357,32 @@ public class FieldNpc : Actor<Npc> {
...
 
-        if (cast is null) {
+    public override SkillRecord? CastSkill(int id, short level, long uid = 0) {
+        if (!Field.SkillMetadata.TryGet(id, level, out SkillMetadata? metadata)) {
+            Logger.Error("Invalid skill use: {SkillId},{Level}", id, level);
             return null;
         }
 
+        if (uid == 0) {
+            uid = (long) ((DateTime.UtcNow - new DateTime(1970, 1, 1)).TotalSeconds * 100000) % (long) 1e14;
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Model/Field/Actor/FieldNpc.cs:368]: nani tf? comment pls or a function to explain

File: Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/TaskState.cs [hunk 10]
```diff
@@ -0,0 +1,148 @@
...
+            if (taskQueue.TryPeek(out NpcTask? task, out NpcTaskPriority priority) && task == pendingTask) {
+                task.Resume();
+            }
+        }
+
+        isPendingStart = false;
+        pendingTask = null;
+    }
+
+    public abstract class NpcTask {
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Model/Field/Actor/ActorStateComponent/TaskState.cs:81]: this needs to be inside TaskState?

File: Maple2.Server.Game/Packets/FieldPacket.cs [hunk 11]
```diff
@@ -213,6 +213,35 @@ public static class FieldPacket {
         return pWriter;
     }
 
+    public static ByteWriter DropDebugItem(FieldItem fieldItem, int objectId, Vector3 position, int unkInt, short unkShort, bool unkBool) {
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Packets/FieldPacket.cs:216]: add a comment explaining what is the purpose of this please

File: Maple2.Server.Game/Packets/NpcControlPacket.cs [hunk 12]
```diff
@@ -33,29 +34,52 @@ public static class NpcControlPacket {
         buffer.WriteByte(2);
 
         buffer.Write<Vector3S>(npc.Position);
-        buffer.WriteShort((short) (npc.Rotation.Z * 10));
-        buffer.Write<Vector3S>(npc.Velocity.Rotate(npc.Rotation));
-        buffer.WriteShort((short) (sequenceSpeed * 100));
+        buffer.WriteShort((short) (npc.Transform.RotationAnglesDegrees.Z * 10));
+        buffer.Write<Vector3S>(npc.MovementState.Velocity);
+        //buffer.Write<Vector3S>(new Vector3(0, 0, 0));
```
Comment (AngeloTadeucci) [inline Maple2.Server.Game/Packets/NpcControlPacket.cs:39]: probably can cleanup some commented code in here

**Review by AngeloTadeucci — CHANGES_REQUESTED** [review]

File: Maple2.Server.Game/Manager/Field/FieldManager.State.cs (hunk 1 above)
Comment (mettaursp) [reply Maple2.Server.Game/Manager/Field/FieldManager.State.cs:177]: Woops, that was for the item spawn debug printing. I put removeme in the name so either I would remove it myself or I'd get barked at in review to remove it if I forgot lol

File: Maple2.Server.Game/Packets/NpcControlPacket.cs (hunk 12 above)
Comment (mettaursp) [reply Maple2.Server.Game/Packets/NpcControlPacket.cs:39]: Woops, I did a pass over for these too but missed this apparently lol

**Review by AngeloTadeucci — APPROVED** [review]

File: Maple2.Server.Game/Manager/Field/FieldManager.State.cs [hunk 13]
```diff
@@ -14,6 +14,7 @@
 using Maple2.Tools;
 using Maple2.Tools.Collision;
 using Maple2.Tools.Extensions;
+using Microsoft.Scripting.Metadata;
```
Comment (Zintixx) [inline Maple2.Server.Game/Manager/Field/FieldManager.State.cs:17]: unused

**Review by Zintixx — APPROVED** [review]

**Pull request merged**